Refactor configuration

mcordell opened the pull request:
Handful of re-factors to clean up the Statistics class.

The major change in this PR is pulling out the configuration of the gem into a dedicate configuration class.  The primary motivation behind this change was to support more flexible backtrace cleaning and logging. The gem user can now set backtrace cleaners and logging via:

```
SqlLogging.configuration.logger = SomeLoggerClass.new
SqlLogging.configuration.backtrace_cleaner = Rails.backtrace_cleaner.dup # Silencer changes
```

The same defaults remain that existed previously but now I can filter the backtrace to include rails engines, further I can write log output to somewhere other that the default logger.

sjmadsen commented on lib/sql-logging/configuration.rb:
```diff
@@ -0,0 +1,42 @@
+    end
+
+    def validate_allowed(allowed_values, value, method_name)
+      return if allowed_values.include?(value)
```
Did you mean to use the `method_name` argument? Because nothing in this method or its one caller actually sets the value if it's valid, and there is no test to assert that it's happening.

mcordell replied:
my mistake, added in fixup

sjmadsen commented on lib/sql-logging/statistics.rb:
```diff
@@ -10,80 +31,87 @@ def configuration
+        return if name.blank? || name =~ / Columns$/ || name == :skip_logging
+        ntuples, bytes = tuples_and_bytes_in_result(result)
+
+        @data.add_query(ntuple, bytes)
```
`ntuple` should be `ntuples`. I know you fixed this typo in the following commit, but I'd've preferred to see the fix squashed into this commit before opening the PR.

sjmadsen commented on test/statistics_test.rb:
```diff
@@ -9,5 +9,22 @@ def test_default_values
+      assert(cleaner.class, Rails::BacktraceCleaner)
+    end
+
+    def test_data
```
You don't introduce the code you're testing here until the next commit. This test should be with it. Again, would've preferred you fixup/squash these kinds of issues before opening a PR.

mcordell replied:
I can rebase this to fix before merge if this PR is going in

sjmadsen commented:
The Travis build is failing across multiple Ruby versions against Rails 4.0. Unless we want to stop supporting 4.0, that must be fixed before merging.

While it won't stop me from merging it, I'd caution against making stylistic changes in a PR. I have nothing against eliminating EOL whitespace, but long lines also don't bother me, and the changes in 9aab7a0 are purely personal preference.

I left some line comments, too.

Finally, while I understand your goal for this PR is likely accomplished, I don't have a clear understanding of _why_ these changes are generally useful. If you could expand in your PR description how the changes benefit other developers and not your specific needs, I'm interested in understanding how the gem falls short for you today.

@tjoyal, thoughts on this PR?

mcordell commented:
>The Travis build is failing across multiple Ruby versions against Rails 4.0. Unless we want to stop supporting 4.0, that must be fixed before merging.

Appears to be a MiniTest version difference, I fixed the tests in question

> While it won't stop me from merging it, I'd caution against making stylistic changes in a PR. I have nothing against eliminating EOL whitespace, but long lines also don't bother me, and the changes in 9aab7a0 are purely personal preference.

Fair enough, I general just work off the ruby style guide

>Finally, while I understand your goal for this PR is likely accomplished, I don't have a clear understanding of why these changes are generally useful. If you could expand in your PR description how the changes benefit other developers and not your specific needs, I'm interested in understanding how the gem falls short for you today.

Updated the description, let me know if it needs anymore clarification.

tjoyal commented on lib/sql-logging/configuration.rb:
```diff
@@ -0,0 +1,47 @@
+
+    def show_top_sql_queries=(value)
+      validate_allowed([false, :rows, :queries, :bytes, :total_time,
+                        :median_time], value, __method__)
```
Thoughts on using `caller_locations` from within the method so it does not have to pollute the call?

mcordell replied:
modified in fixup

tjoyal commented on lib/sql-logging/configuration.rb:
```diff
@@ -0,0 +1,47 @@
+
+    def initialize
+      DEFAULTS.each do |k, v|
+        instance_variable_set("@#{k}", v)
```
Why not go through the accessors where extra validation is/might be performed?

mcordell replied:
modified in fixup

tjoyal commented on lib/sql-logging/configuration.rb:
```diff
@@ -0,0 +1,47 @@
+      logger: nil
+    }.freeze
+
+    attr_accessor(*DEFAULTS.keys)
```
On a personal note I would have defined 2 methods for each of these since we end up with 3 out of 10 that gets overridden, but I'm also ok keeping it that way.

mcordell replied:
modified in fixup

tjoyal commented on lib/sql-logging/configuration.rb:
```diff
@@ -0,0 +1,47 @@
+      show_sql_backtrace: true,
+      show_top_sql_queries: :total_time,
+      top_sql_queries: 10,
+      backtrace_cleaner: nil,
```
Let's reference these new configs in the README

tjoyal commented on lib/sql-logging/configuration.rb:
```diff
@@ -0,0 +1,47 @@
+
+    def initialize_cleaner
+      Rails.backtrace_cleaner.dup.tap do |cleaner|
+        cleaner.add_silencer { |line| line =~ %r{sql-logging/lib} }
```
Should setting the `backtrace_cleaner` lead to this line to be appended in some way so the user does not have to keep this on their end?

tjoyal commented on lib/sql-logging/statistics.rb:
```diff
@@ -1,105 +1,119 @@
+        SqlLogging.configuration
+      end
+
+      def_delegators :configuration, :show_sql_backtrace, :top_sql_queries,
```
I feel we should implement the breaking change(remove), or keep these with a deprecation notice when called.

tjoyal replied:
That PR is already large. We could split it differently, but I'm not against keeping it in this format and iterate. We can keep this and discuss it afterward.

tjoyal commented:
I like where this is going. Somewhat large, could have been split between code cleanup and feature change, but it's all good for me if we want to do it at once.

Splitting code in multiple classes really makes me happy.

I like to do breaking changes as I go to not have to maintain old code paths(`def_delegators`) and have a change log so it is easy to upgrade. Not sure what you think about that @sjmadsen ?

mcordell commented:
Added README updates and fix ups for the additional comments. Let me know of any other fixes needed to get this merged.

I didn't do anything about the deprecation idea because I figured that was a decision better left to you two.

sjmadsen commented:
Hang tight on this for a bit more, Michael. This is a busy week for both Thierry and I. Should get back to it this coming weekend or early next week.

tjoyal commented:
Very sorry for the long delay, but I'm back. Let's try to get this out of the door!

I see you added the new configs in the README, and I really like the idea of a configuration object. Could we update the "Options" section to reflect the new way of configuring it?

tjoyal commented:
And if we could get some simple test coverage for `SqlLogging::Configuration` that would be great. Especially around those entry points that use the private methods logic.

mcordell commented:
> I see you added the new configs in the README, and I really like the idea of a configuration object. Could we update the "Options" section to reflect the new way of configuring it?

I'm not sure I follow what you mean here?

Added a commit that adds coverage for the configuration class

mcordell commented:
Just checking back in, what is the status of this PR? Do I need to fix anything else?

tjoyal commented:
https://github.com/lightyear/sql-logging/blob/master/README.rdoc#options

Update it to use `SqlLogging.configuration` and I'll be good with it.

We can take care of deprecating these old `def_delegators` in some subsequent work.

mcordell commented:
@tjoyal I updated the configuration as requested and re-based this branch. Thanks!

tjoyal commented:
👍 Let's squash it and we'll go with all of it!

mcordell commented:
Do you want me to squash it? I rebased it to clean up the commit history.

tjoyal commented:
> Do you want me to squash it? I rebased it to clean up the commit history.

Missed that part. Let's get it into the trunk!
